src/transport/CertificateAuthenticatedSessionEstablishment.cpp

```diff
@@ -0,0 +1,415 @@
+
+    mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaErr;
+
+    VerifyOrExit(header.GetSourceNodeId().ValueOr(kUndefinedNodeId) == mConnectionState.GetPeerNodeId(),
+                 err = CHIP_ERROR_WRONG_NODE_ID);
+    VerifyOrExit(header.GetEncryptionKeyID() == mConnectionState.GetPeerKeyID(), err = CHIP_ERROR_INVALID_KEY_ID);
+
+    mPairingComplete = true;
+
+    // Call delegate to indicate pairing completion
+    mDelegate->OnPairingComplete();
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(SigmaErrorType::kUnexpected);
+    }
+    return err;
+}
+
+void CertificateAuthenticatedSessionEstablishment::SendErrorMsg(SigmaErrorType errorCode)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    System::PacketBufferHandle msg;
+    uint16_t msglen      = sizeof(SigmaErrorMsg);
+    SigmaErrorMsg * pMsg = nullptr;
+
+    msg = System::PacketBuffer::NewWithAvailableSize(msglen);
+    VerifyOrExit(!msg.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
+
+    pMsg        = reinterpret_cast<SigmaErrorMsg *>(msg->Start());
+    pMsg->error = errorCode;
+
+    msg->SetDataLength(msglen);
+
+    err = AttachHeaderAndSend(Protocols::SecureChannel::MsgType::CASE_SigmaErr, std::move(msg));
+    SuccessOrExit(err);
+
+exit:
+    Clear();
+}
+
+void CertificateAuthenticatedSessionEstablishment::HandleErrorMsg(const PacketHeader & header,
+                                                                  const System::PacketBufferHandle & msg)
+{
+    // Error message processing
+    const uint8_t * buf  = msg->Start();
+    size_t buflen        = msg->DataLength();
+    SigmaErrorMsg * pMsg = nullptr;
+
+    VerifyOrExit(buf != nullptr, ChipLogError(Inet, "Null error msg received during pairing"));
+    VerifyOrExit(buflen == sizeof(SigmaErrorMsg), ChipLogError(Inet, "Error msg with incorrect length received during pairing"));
+
+    pMsg = reinterpret_cast<SigmaErrorMsg *>(msg->Start());
+    ChipLogError(Inet, "Received error (%d) during CASE pairing process", pMsg->error);
+
+exit:
+    Clear();
+}
+
+CHIP_ERROR CertificateAuthenticatedSessionEstablishment::HandlePeerMessage(const PacketHeader & packetHeader,
+                                                                           const Transport::PeerAddress & peerAddress,
+                                                                           System::PacketBufferHandle msg)
+{
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint16_t headerSize = 0;
+    PayloadHeader payloadHeader;
+
+    VerifyOrExit(!msg.IsNull(), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    err = payloadHeader.Decode(msg->Start(), msg->DataLength(), &headerSize);
+    SuccessOrExit(err);
+
+    msg->ConsumeHead(headerSize);
+
+    VerifyOrExit(payloadHeader.GetProtocolID() == Protocols::kProtocol_SecureChannel, err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
+    VerifyOrExit(payloadHeader.GetMessageType() == (uint8_t) mNextExpectedMsg, err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
```

Comment:
Since right below we end up casting the message type to the right type anyway, how about we do it before this line and avoid the cast here?